osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs

```diff
@@ -123,13 +127,41 @@ public override bool HandleFlip(Direction direction, bool flipOverOrigin)
         {
             var hitObjects = selectedMovableObjects;
 
-            var flipQuad = flipOverOrigin ? new Quad(0, 0, OsuPlayfield.BASE_SIZE.X, OsuPlayfield.BASE_SIZE.Y) : GeometryUtils.GetSurroundingQuad(hitObjects);
+            // If we're flipping over the origin, we take the grid origin position from the grid toolbox.
+            var flipQuad = flipOverOrigin ? new Quad(gridToolbox.StartPositionX.Value, gridToolbox.StartPositionY.Value, 0, 0) : GeometryUtils.GetSurroundingQuad(hitObjects);
+            Vector2 flipAxis = direction == Direction.Vertical ? Vector2.UnitY : Vector2.UnitX;
+
+            if (flipOverOrigin)
+            {
+                // If we're flipping over the origin, we take one of the axes of the grid.
+                // Take the axis closest to the direction we want to flip over.
+                switch (gridToolbox.GridType.Value)
+                {
+                    case PositionSnapGridType.Square:
+                        flipAxis = GeometryUtils.RotateVector(Vector2.UnitX, -((gridToolbox.GridLinesRotation.Value + 405) % 90 - 45));
+                        flipAxis = direction == Direction.Vertical ? flipAxis.PerpendicularLeft : flipAxis;
+                        break;
+
+                    case PositionSnapGridType.Triangle:
+                        // Hex grid has 3 axes, so you can not directly flip over one of the axes,
+                        // however it's still possible to achieve that flip by combining multiple flips over the other axes.
+                        flipAxis = direction == Direction.Vertical
+                            ? GeometryUtils.RotateVector(Vector2.UnitX, -((gridToolbox.GridLinesRotation.Value + 390) % 60 + 60))
```

Comment:
I think you misread the position of the unary minus operator. It takes place after the addition because the minus is outside the parentheses. The `-((gridToolbox.GridLinesRotation.Value + 390) % 60 + 60)` is in the range [-120;-60], so thats a 90 degree rotation of the range of the other flip axis which is in [-30;30].

Yes, the rotation slider is constrained to [-30;30], but this part of the code considers flipping over any of the 3 axes that exist on the triangle grid, and they are at multiples of 60 degrees rotation from the base rotation, so the chosen flip axis is not necessarily in [-30;30]. Also note that the flip axis vector is perpendicular to the axis we're flipping/mirroring over.